src/webgpu/web_platform/copyToTexture/image.spec.ts

```diff
@@ -339,3 +346,113 @@ g.test('copy_subrect_from_2D_Canvas')
       { maxDiffULPsForFloatFormat: 1, maxDiffULPsForNormFormat: 1 }
     );
   });
+
+g.test('from_rotated_image')
+  .desc(
+    `
+    Test HTMLImageElements with rotation metadata can be copied to WebGPU texture correctly.
+
+    It creates Images with images under Resource folder.
+
+    Then call copyExternalImageToTexture() to do a full copy to the 0 mipLevel
+    of dst texture, and read one pixel out to compare with the manually documented expected color.
+
+    If 'flipY' in 'GPUCopyExternalImageSourceInfo' is set to 'true', copy will ensure the result
+    is flipped.
+
+    The tests covers:
+    - Image with rotation metadata
+    - Valid 'flipY' config in 'GPUCopyExternalImageSourceInfo' (named 'srcDoFlipYDuringCopy' in cases)
+    - TODO: partial copy tests should be added
+    - TODO: all valid dstColorFormat tests should be added.
+  `
+  )
+  .params(u =>
+    u //
+      .combine('imageName', kImageNames)
+      .combine('srcDoFlipYDuringCopy', [true, false])
+  )
+  .fn(async t => {
+    const { imageName, srcDoFlipYDuringCopy } = t.params;
+    const kColorFormat = 'rgba8unorm';
+
+    // Load image.
+    const image = new Image();
+    const imageUrl = getResourcePath(imageName);
+    image.src = imageUrl;
+    await raceWithRejectOnTimeout(image.decode(), 5000, 'decode image timeout');
+    const width = image.width;
+    const height = image.height;
+
+    const dstTexture = t.createTextureTracked({
+      size: { width, height },
+      format: kColorFormat,
+      usage:
+        GPUTextureUsage.COPY_DST | GPUTextureUsage.COPY_SRC | GPUTextureUsage.RENDER_ATTACHMENT,
+    });
+
+    t.device.queue.copyExternalImageToTexture(
+      {
+        source: image,
```

Comment:
It may simplify the tests to just have a parameter for how we upload the image, instead of having two very similar tests - because the rest of the test (the checking code) will be the same. (There's also at least one more code path that it would be good to test, which is ImageBitmap from blob.)

(If you do that then maybe it should go in a new file like `copyToTexture/image_file.spec.ts` since it's not divided up the same way as the other tests.)

To track writing more tests like that I've filed #4114. If you think of any other upload methods please log them on that issue. If you don't cover all of those in this PR, feel free to leave a TODO for now.

Reply:
Done.